backend/src/controllers/regionController.js

```diff
@@ -1,56 +1,146 @@
 
+async function getAllSubregions(regionId, hierarchyId) {
+  const query = `
+        WITH RECURSIVE Subregions AS (
+            SELECT *
+            FROM hierarchy
+            WHERE parent_id = :regionId AND hierarchy_id = :hierarchyId
+            UNION ALL
+            SELECT h.*
+            FROM hierarchy h
+            INNER JOIN Subregions s ON h.parent_id = s.region_id AND h.hierarchy_id = :hierarchyId
+        )
+        SELECT * FROM Subregions;
+    `;
 
-exports.getHierarchies = async (req, res) => {
-    try {
-        const hierarchies = await HierarchyNames.findAll();
-        res.status(200).json(hierarchies.map(hierarchy => hierarchy.toApiFormat()));
-    } catch (err) {
-        console.error(err);
-        res.status(500).json({ message: 'Internal Server Error' });
-    }
+  return sequelize.query(query, {
+    replacements: { regionId, hierarchyId },
+    type: QueryTypes.SELECT,
+    mapToModel: true,
+    model: Hierarchy,
+  });
 }
 
-exports.getGeometry = async (req, res) => {
-    const regionId  = req.params.regionId;
-    const resolveEmpty = req.query.resolveEmpty === 'true';
-    const hierarchyId = req.query.hierarchyId || 1;
-
+// Retrieve subregions for a specific region
+async function getSubregions(regionId, hierarchyId, getAll) {
+  try {
     // Check if the region exists
     const region = await Hierarchy.findOne({
-        where: {
-            regionId: regionId,
-            hierarchyId: hierarchyId
-        },
-    })
+      where: {
+        regionId,
+        hierarchyId,
+      },
+    });
+
     if (!region) {
-        return res.status(404).json({ message: 'Region not found' });
+      return { data: [], message: 'Region not found', status: 404 };
+    }
+
+    // Retrieve subregions
+    let subregions;
+    // Check the getAll query parameter
+    if (getAll === 'true') {
+      subregions = await getAllSubregions(regionId, hierarchyId);
+    } else {
+      subregions = await Hierarchy.findAll({
+        where: { parentId: regionId, hierarchyId },
+        mapToModel: true,
+        model: Hierarchy,
+      });
+    }
+
+    if (subregions.length === 0) {
+      return { data: [], message: 'Region has no subregions', status: 202 };
     }
 
-    //Find all subdivisions of the region
-    const divisions = await getDivisions(regionId, hierarchyId, res);
-    let geometries = [];
-    let notCompleted = false;
-    for (let division of divisions) {
-        const regionId = division.id;
-        const geom = division.geom;
-        if (geom) {
-            geometries.push(geom);
-        } else {
-            if (!resolveEmpty) {
-                notCompleted = true;
-                break;
-            } else {
-                // Find ids and geometries of all subdivisions of the division
-                const query = `
+    return { data: subregions, status: 200 };
+  } catch (err) {
+    console.error(err);
+    return { data: [], message: 'Internal Server Error', status: 500 };
```

Comment:
**Suggestion:** Use a more descriptive variable name than 'err' for better readability.
```suggestion
    console.error(error);
    return { data: [], message: 'Internal Server Error', status: 500 };
```